File: RUET IUPC 2019/Solutions/A.cpp
/* BISMILLAHIR RAHMANIR RAHIM */
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;

const int mxn = 1e5 + 5;

class ChineseReminderTheorm {
public:
    vector<ll> A, M;

    void clear() {
        A.clear(), M.clear();
    }

    ll extendedEuclid( ll a, ll b, ll& x, ll& y ) {
        if( a == 0 ) {
            x = 0, y = 1;
            return b;
        }
        ll x1, y1;
        ll gcd = extendedEuclid( b % a, a, x1, y1 );

        x = y1 - (b / a) * x1;
        y = x1;
        return gcd;
    }

    pair<ll, ll> solve() {
        if( A.empty() or A.size() != M.size() ) return {-1, -1}; // invalid input

        ll a1 = A[0];
        ll m1 = M[0];
        a1 %= m1;

        for(int i = 1; i < A.size(); i++) {
            ll a2 = A[i];
            ll m2 = M[i];

            ll gcd = __gcd(m1, m2);
            if( a1 % gcd != a2 % gcd ) return {-1, -1}; // conflict in equation

            ll p, q;
            extendedEuclid( m1 / gcd, m2 / gcd, p, q );

            ll mod = (m1 / gcd) * m2;
            ll tmp = ((__int128)a1 * (m2 / gcd) * q + (__int128)a2 * (m1 / gcd) * p) % mod;
            a1 = tmp;
            if( a1 < 0 ) a1 += mod;
            m1 = mod;
        }
        return {a1, m1};
    }
};

int main() {
    ios_base :: sync_with_stdio(0);
    cin.tie(0), cout.tie(0);

	vector<int> primes;
	vector<bool> vis(mxn, false);
    for(int i = 2; i < mxn; i++) {
        if( !vis[i] ) {
            primes.push_back(i);
            for(int j = i + i; j < mxn; vis[j] = true, j += i);
        }
    }

	vector<int> factors[mxn];
    for(int i = 1; i < mxn; i++) {
        int n = i;
        for(auto& p : primes) {
            if( p > n ) break;
            int tmp = 1;
            while( n > 0 and n % p == 0 ) {
                n /= p, tmp *= p;
            }
            if( tmp == 1 ) continue;
            factors[i].push_back(tmp);
        }
        if( n > 1 ) factors[i].push_back(n);
    }

    int cases, caseno = 0;
    cin >> cases;
    while( cases-- ) {
        string N;
        int M;
        cin >> N >> M;

        auto findMod = [&](int x) {
            ll rem = 0, lim = 1e17;
            for(auto d : N) {
                rem = (rem * 10) + (d - '0');
                if( rem >= lim ) {
                    rem %= x;
                }
            }
            return (int)(rem % x);
        };

        vector<int> res(M + 2, -1);
        res[0] = res[1] = 0;
        for(auto p : primes) {
            if( p > M ) break;
            ll maxP = p;
            while( p * maxP <= M ) {
            	maxP *= p;
            }
            res[maxP] = findMod( maxP );
            for(int i = maxP / p; i >= p; i /= p) {
                res[i] = res[maxP] % i;
            }
        }

        ChineseReminderTheorm crt;
        for(int i = M; i > 1; i--) {
            if( ~res[i] ) continue;
            if( i << 1 <= M ) res[i] = res[i << 1] % i;
            else {
                crt.clear();
                for(auto& f : factors[i]) {
                    crt.M.push_back(f);
                    crt.A.push_back(res[f]);
                }
                res[i] = crt.solve().first;
            }
        }

        ll ans = 0;
        for(int i = 1; i <= M; i++) {
            ans += res[i];
        }
        cout << "Case " << ++caseno << ": " << ans << "\n";
    }
}

File: RUET IUPC 2019/Solutions/A.py
"""BISMILLAHIR RAHMANIR RAHIM"""

import math

mxn = 100005
vis = [False] * mxn
primes = []
factors = [[]]


def sieve():
    vis[0] = vis[1] = True
    for i in range(2, mxn):
        if not vis[i]:
            primes.append(i)
            for j in range(i + i, mxn, i):
                vis[j] = True


def prime_factorize():
    for i in range(1, mxn):
        n = i
        factors.append([])
        for p in primes:
            if p > n:
                break
            tmp = 1
            while n > 0 and n % p == 0:
                n /= p
                tmp *= p

            if tmp == 1:
                continue
            factors[i].append(tmp)

        if n > 1:
            factors[i].append(n)


def extended_euclid(a, b):
    if a == 0:
        return b, 0, 1
    else:
        gcd, x, y = extended_euclid(b % a, a)
        return gcd, y - (b // a) * x, x


def chinese_reminder_theorem(A, P):
    if len(A) == 0 or len(A) != len(P):
        return 0, 0

    a1 = A[0]
    m1 = P[0]
    for i in range(1, len(A)):
        a2 = A[i]
        m2 = P[i]
        g = math.gcd(m1, m2)
        if a1 % g != a2 % g:
            return -1, -1
        gcd, p, q = extended_euclid(m1 // g, m2 // g)
        mod = (m1 // g) * m2
        x = (a1 * (m2 // g) * q + a2 * (m1 // g) * p) % mod
        a1 = x
        if a1 < 0:
            a1 += mod
        m1 = mod

    return a1, m1


def main():
    sieve()
    prime_factorize()

    cases = int(input())
    for case_no in range(1, cases + 1):
        N, M = map(int, input().split())
        res = [-1] * (M + 1)
        res[0] = res[1] = 0
        for p in primes:
            if p > M:
                break
            max_p = 1
            while max_p * p <= M:
                max_p *= p

            res[max_p] = N % max_p
            i = max_p // p
            while i >= p:
                res[i] = res[max_p] % i
                i //= p

        for i in range(M, 1, -1):
            if res[i] != -1:
                continue
            if i * 2 <= M:
                res[i] = res[i * 2] % i
            else:
                P = factors[i]
                A = []
                for fct in factors[i]:
                    A.append(res[fct])
                tmp_x, tmp_m = chinese_reminder_theorem(A, P)
                res[i] = tmp_x

        ans = sum(res)
        print('Case %d: %d' % (case_no, ans))


main()